include/boost/url/bnf/parse.hpp

```diff
@@ -77,214 +96,182 @@ parse(
-       ec = error::success;
-       return true;
-     }
-   };
-   } // namespace detail
-   } // namespace urls
-   } // namespace boost
-
-   int main() {
-     using namespace boost::urls;
-
-     // Input
-     std::string str = "Hello.cpp";
-     auto it = str.data();
-     auto const end = it + str.size();
-
-     // Output
-     std::string alpha_l;
-     std::string alpha_r;
-     error_code ec;
-
-     // Parse two alpha strings separated by a '.'
-     if (bnf::parse(it, end, ec, detail::alpha_string_bnf{alpha_l}, '.',
-                    detail::alpha_string_bnf{alpha_r})) {
-       std::cout << alpha_l << std::endl;
-       std::cout << alpha_r << std::endl;
-     }
-     return 0;
+   if (bnf::parse(it, end, ec, bnf_element_1, bnf_element_2, bnf_element_3)) {
+     std::cout << "Range [it, end) parsed successfully\n";
    }
    @endcode
 
    @par Exception Safety
-     Calls to `parse` overloads may throw, although they should
-     prefer setting @ref ec instead.
 
-   @see https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form
+     Internal calls to `parse` overloads for `args` may throw, although they should
+     prefer setting `ec` instead.
+
+   @see
+
+   [Backus–Naur form](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form),
 
-        @ref boost::urls::error
+   @ref boost::urls::error
 
    @return `true` if the range initial chars match the elements successfully,
    `false` otherwise
 
-   @param[in,out] it An iterator to the first element in the range.
-   At the end of this function, @ref it points to one past the last
+   @param it An iterator to the first element in the range.
+   At the end of this function, `it` points to one past the last
    element parsed in the range.
 
-   @param[in] end An iterator to one past the last element in the range
+   @param end An iterator to one past the last element in the range
 
-   @param[out] ec The error code this function should set if it fails
+   @param ec The error code this function should set `it` it fails
 
-   @param[out] args BNF elements to be parsed
+   @param args BNF elements to be parsed
 
 */
+template<
 #ifdef BOOST_URL_DOCS
-template<class... Args>
-bool parse( char const*& it, char const* end, Args&&... args);
+    class... Args
 #else
-template<
     class T0,
     class T1,
-    class... Tn>
+    class... Tn
+#endif
+>
 bool
 parse(
     char const*& it,
     char const* end,
     error_code& ec,
+#ifdef BOOST_URL_DOCS
+    Args&&... args
+#else
     T0&& t0,
     T1&& t1,
-    Tn&&... tn);
+    Tn&&... tn
 #endif
+);
 
-/** @brief Parse a complete string
+/** Parse a complete string as a sequence of BNF elements
 
-   This function parses a complete string @ref s into the Backus–Naur form (BNF)
-   elements [@ref t0, @ref t1, ...]. If the string is not completely consumed,
-   @ref ec is set to @ref error::leftover_input.
+   This function parses a complete string `s` into the Backus–Naur form (BNF)
+   elements [`arg_0`, `arg_1`, ...] in `args`.
```

Comment:
The important thing you need to state is "the function returns an error if some characters are left unparsed"

Reply:
OK. I'm adding that. That differentiation was initially in the implicit `brief`.

Reply:
Its hard to say it concisely in the brief